### DISCUSSION TITLE
REST API search with float16

### DISCUSSION THREAD

**Discussion by yifan** [milvus-io/milvus#37123] [Q&A and General discussion]
how to search rest api with float16 index and float 16 embedding? I am using with a list of float

```
fail to search on QueryNode 1: worker(1) query failed: 
Assert \"static_cast<int>(field_meta.get_data_type()) == static_cast<int>(info.type())\" at 
/workspace/source/internal/core/src/query/Plan.cpp:48\n => 
vector type must be the same, field vector - type VECTOR_FLOAT16, search info type VECTOR_FLOAT
```

**Comment by yhmo** [comment]
Are you using python to call the rest api?

**Reply by xiaofan-luan** [reply]
should we just convert FP32 to fp16 automatically?

@smellthemoon @liliu-z @tedxu any thoughts on this?

**Reply by yifan** [reply]
> Are you using python to call the rest api?

Yes, I am using the REST api, because pymilvus async doesn't work, it is better for us to use the REST API directly. Because JSON does not really have FLOAT16 or FLOAT32, it is just float, not sure why it is not converted automatically to FLOAT16 to match the index.

**Reply by xiaofan-luan** [reply]
what do you mean by " pymilvus async " is not working? can you give an example?

**Reply by xiaofan-luan** [reply]
And i don't understand what make restful API different becasue restful is always gonna to be access synchrously

**Reply by yifan** [reply]
it is mainly from the client, not from server's point of view. On client side, async http request means I can do await and handle multiple requests in my code. pymilvus client has async mode can be enabled by _async argument, however there are multiple issues on this:
https://github.com/milvus-io/pymilvus/issues/2198

Regarding on the original question, I would expect when converting JSON to native datatypes, floats will be converted into the same precision as the index.

**Reply by yifan** [reply]
one question regarding FLOAT16_VECTOR, is it supported natively in Milvus, and it should reduce the memory usage by half?

**Reply by xiaofan-luan** [reply]
Yes, I think for latest release, the FP16 can save you half memories.
Older versions might not, since FP is simulated by FP32
@liliu-z

**Reply by liliu-z** [reply]
> should we just convert FP32 to fp16 automatically?
> 
> @smellthemoon @liliu-z @tedxu any thoughts on this?

This is worth debating:
If yes, means we modify the data from users' side and store it. Some operations like `GetVector`'s meaning will get hurt.
If no, SDK user experience, especially RESTFUL API will suffer

**Reply by liliu-z** [reply]
> Yes, I think for latest release, the FP16 can save you half memories. Older versions might not, since FP is simulated by FP32 @liliu-z

@yifan Depends on your index. Native FP/BF16 support for HNSW & DiskANN are ready from 2.4.

**Reply by xiaofan-luan** [reply]
> > should we just convert FP32 to fp16 automatically?
> > @smellthemoon @liliu-z @tedxu any thoughts on this?
> 
> This is worth debating: If yes, means we modify the data from users' side and store it. Some operations like `GetVector`'s meaning will get hurt. If no, SDK user experience, especially RESTFUL API will suffer

They define their vector as BF16/FP16 already. and there is no easy way to represent BF16/FP16 on most the languages. If data lose accuracy it's user's choice.

If they want to keep original data they should use Fp32 float + Floa16/BF16 quantazition or int8 quantazation in the future

**Reply by liliu-z** [reply]
> > > should we just convert FP32 to fp16 automatically?
> > > @smellthemoon @liliu-z @tedxu any thoughts on this?
> > 
> > 
> > This is worth debating: If yes, means we modify the data from users' side and store it. Some operations like `GetVector`'s meaning will get hurt. If no, SDK user experience, especially RESTFUL API will suffer
> 
> They define their vector as BF16/FP16 already. and there is no easy way to represent BF16/FP16 on most the languages. If data lose accuracy it's user's choice.
> 
> If they want to keep original data they should use Fp32 float + Floa16/BF16 quantazition or int8 quantazation in the future

Make sense, let's plan this next release
